File: javascript/space-age/space-age.js
const SECONDS_OF_YEAR = 31557600

const RATIOS = Object.freeze({
  earth: 1,
  mercury: 0.2408467,
  venus: 0.61519726,
  mars: 1.8808158,
  jupiter: 11.862615,
  saturn: 29.447498,
  uranus: 84.016846,
  neptune: 164.79132,
})

const makeAgeOn = seconds => planet => {
  if (! RATIOS.hasOwnProperty(planet)) {
    throw new Error(`The planet ${planet} does not exist`)
  }
  const age = seconds / (SECONDS_OF_YEAR * RATIOS[planet])
  return Math.round(age * 100) / 100
}

module.exports = class SpaceAge {
  constructor(seconds) {
    this.seconds = seconds
    this.getAgeOn = makeAgeOn(seconds)
  }

  onEarth() {
    return this.getAgeOn('earth')
  }

  onMercury() {
    return this.getAgeOn('mercury')
  }

  onVenus() {
    return this.getAgeOn('venus')
  }

  onMars() {
    return this.getAgeOn('mars')
  }

  onJupiter() {
    return this.getAgeOn('jupiter')
  }

  onSaturn() {
    return this.getAgeOn('saturn')
  }

  onUranus() {
    return this.getAgeOn('uranus')
  }

  onNeptune() {
    return this.getAgeOn('neptune')
  }
}

File: go/space-age/space_age.go
package space

type Planet string

const EARTH_YEAR_SEC = 31557600

func Age(sec float64, planet Planet) float64 {
	rate := getYearRate(planet)
	return sec / (EARTH_YEAR_SEC * rate)
}

func getYearRate(planet Planet) float64 {
	switch planet {
	case "Mercury":
		return 0.2408467
	case "Venus":
		return 0.61519726
	case "Mars":
		return 1.8808158
	case "Jupiter":
		return 11.862615
	case "Saturn":
		return 29.447498
	case "Uranus":
		return 84.016846
	case "Neptune":
		return 164.79132
	default:
		return 1
	}
}
